### ISSUE TITLE
saving data (though not updated entries) with whitelist renders wrong sql

### ISSUE THREAD

**Issue by jovanialferez** [UnionOfRAD/lithium#1163]
i had this snippet which was working as of 6eecce29:

$model->fieldOne = 'same_value_as_before';
$model->save(null, ['whitellist' => ['fieldOne']]);

If there were no actual update, such that fieldOne is still assigned to same value, save() causes to render an sql with no field SET e.g. `update my_table SET WHERE id=123`.

**Comment by jovanialferez** [comment]
*8e5f8a36bdf6db2aa3ccf3133f2fed2b6eecce29

**Comment by davidpersson** [comment]
I think this is related to 00f80a2080c1c11a66f27b88c42c4b05d2674ca4. @koushki do you think there is an easy fix for that?

**Comment by koushki** [comment]
@davidpersson yes I'll fix that as soon as possible.

**Comment by farhadi** [comment]
The default behavior, before @koushki's patch, was to do a full replace which could be problematic in cases where the developer expects an update only on some specific fields that have been changed.
But in the current behavior, modified fields will be updated by comparing the current state to the initial state, and in cases where there is no field to be updated the model's key would be placed in the SET fragment to avoid rendering an invalid sql, and in case with a whitelist not containing model's key, the problem mentioned in this issue would happen.

I think both approaches (i.e. before and after koushki's patch) for saving an entity are logically somehow incorrect.
IMO the best behavior is to update only the fields which have been explicitly assigned a value no matter whether the field is changed or not, and in cases where there is nothing to be updated, the `save` should return without executing a sql.

**Comment by davidpersson** [comment]
I have reverted the original commit that added the incremental update feature to the `dev` branch as - after some investigation - I didn't find a fix by myself for this. This feature probably needs a little bit more work (see @farhadi's comment) and we're fine.  

However I really would like to see this feature in 1.0.

Therefore I've created a feature branch under `feature/incremental-db-update`. That contains the original patch and test cases by @koushki as well as some tests added by me.

@koushki if your or anybody else would like to pick up the work here, take the commits from that branch and submit a new PR, I'd be glad to help.
